### PR TITLE
[s] Fixes Changeling Digital Camo with AI

### LINKED ISSUE #36759: Changeling Digital Camo Exploit
Tested with the current repo.

When changeling uses digital camo, it goes invisible to the AI. But when the changeling toggles off the digital camo, he is still invisible to the AI. Also #22040 is still broken.

1. Become an AI as your 1st client
2. Become a changeling as your 2nd client
3. Get digital camo upgrade from changeling
4. Turn it on, you'll see the changeling go invisible with your AI client.
5. Turn it off, the AI client will still see the changeling client as invisible.

### PR DESCRIPTION
**Pull request by Davidj361**
[Changelogs]:

:cl:
fix: AI can now see changeling when changeling disables its digital camo.
/:cl:

Fixes #36759

### CODE REVIEW

File: code/game/alternate_appearance.dm [hunk 1]
```diff
@@ -113,6 +113,19 @@ GLOBAL_LIST_EMPTY(active_alternate_appearances)
 		return TRUE
 	return FALSE
 
+/datum/atom_hud/alternate_appearance/basic/AI
+
+/datum/atom_hud/alternate_appearance/basic/AI/New()
+	..()
+	for(var/mob/living/silicon/ai/AI in GLOB.silicon_mobs)
```
Comment (Cyberboss) [inline code/game/alternate_appearance.dm]: we have GLOB.ai_list I believe. If so, use it and make the loop typeless

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/alternate_appearance.dm [hunk 2]
```diff
@@ -113,6 +113,19 @@ GLOBAL_LIST_EMPTY(active_alternate_appearances)
 		return TRUE
 	return FALSE
 
+/datum/atom_hud/alternate_appearance/basic/AI
+
+/datum/atom_hud/alternate_appearance/basic/AI/New()
+	..()
+	for(var/mob/living/silicon/ai/AI in GLOB.ai_list)
+		if(mobShouldSee(AI))
```
Comment (Cyberboss) [inline code/game/alternate_appearance.dm]: make the loop typeless

Comment (Cyberboss) [reply code/game/alternate_appearance.dm]: Also since your `mobShouldSee` just checks if the mob is an AI, the call to it is unecessary

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/alternate_appearance.dm (hunk 2 above)
Comment (Davidj361) [reply code/game/alternate_appearance.dm]: Right, I only copied the format of the older procs.

**Comment by Davidj361** [comment]
I'm going to do some code changes and "refactor". Don't merge yet.

**Comment by Davidj361** [comment]
Ok it should be good. I removed the conditionals of If(...) for shouldsee and replaced that functionality with GLOB lists.

File: code/game/alternate_appearance.dm [hunk 3]
```diff
@@ -93,63 +93,44 @@ GLOBAL_LIST_EMPTY(active_alternate_appearances)
 
 /datum/atom_hud/alternate_appearance/basic/everyone/New()
 	..()
-	for(var/mob in GLOB.mob_list)
-		if(mobShouldSee(mob))
-			add_hud_to(mob)
-
-/datum/atom_hud/alternate_appearance/basic/everyone/mobShouldSee(mob/M)
-	return !isobserver(M)
+	for(var/mob in (GLOB.mob_list - GLOB.dead_mob_list))
```
Comment (Cyberboss) [inline code/game/alternate_appearance.dm]: GLOB.alive_mob_list

File: code/game/alternate_appearance.dm [hunk 4]
```diff
@@ -93,63 +93,44 @@ GLOBAL_LIST_EMPTY(active_alternate_appearances)
...
 
-/datum/atom_hud/alternate_appearance/basic/silicons/mobShouldSee(mob/M)
-	if(issilicon(M))
-		return TRUE
-	return FALSE
+/datum/atom_hud/alternate_appearance/basic/AI
+
+/datum/atom_hud/alternate_appearance/basic/AI/New()
+	..()
+	for(var/mob/ai in GLOB.ai_list)
```
Comment (Cyberboss) [inline code/game/alternate_appearance.dm]: typeless

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/alternate_appearance.dm (hunk 4 above)
Comment (Davidj361) [reply code/game/alternate_appearance.dm]: Ah sorry, I thought it was.

**Comment by Davidj361** [comment]
@Cyberboss Can you clarify this?
I don't know if doing list A - list B is a more costly operation than iterating through a list of both and filtering using a conditional. So clarify that for the noncultist part.

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
A - B is fine. Fix travis

**Comment by ninjanomnom** [comment]
Ping for reopen if you get back to working on this

**Comment by Davidj361** [comment]
Merged this branch with #36754

**Pull request closed without merging**